src/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -53,101 +106,173 @@ private static Delegate CreateObjectArrayDelegateRefEmit(Type delegateType, Func
+                    if (hasReturnValue)
+                        thunkTypeArgCount++;
 
-                    ilgen.Emit(OpCodes.Ldloc, argArray);
-                    ilgen.Emit(OpCodes.Ldc_I4, i);
-                    ilgen.Emit(OpCodes.Ldarg, i + 1);
+                    Type[] thunkTypeArgs = new Type[thunkTypeArgCount];
+                    for (int i = 0; i < parameters.Length; i++)
+                    {
+                        thunkTypeArgs[i] = parameters[i].ParameterType;
+                    }
+                    MethodInfo uninstantiatedMethod;
 
-                    if (paramIsByReference)
+                    if (hasReturnValue)
                     {
-                        ilgen.Emit(OpCodes.Ldobj, paramType);
+                        thunkTypeArgs[thunkTypeArgs.Length - 1] = returnType;
+                        uninstantiatedMethod = s_FuncThunks[parameters.Length];
+                    }
+                    else
+                    {
+                        uninstantiatedMethod = s_ActionThunks[parameters.Length];
                     }
-                    Type boxType = ConvertToBoxableType(paramType);
-                    ilgen.Emit(OpCodes.Box, boxType);
-                    ilgen.Emit(OpCodes.Stelem_Ref);
-                }
 
-                if (hasRefArgs)
+                    if (thunkTypeArgs.Length > 0)
+                        s_thunks[delegateType] = thunkMethod = uninstantiatedMethod.MakeGenericMethod(thunkTypeArgs);
+                    else
+                        s_thunks[delegateType] = thunkMethod = uninstantiatedMethod;
+
+                } while (false);
+
+                if (thunkMethod == null)
                 {
-                    ilgen.BeginExceptionBlock();
-                }
+                    int thunkIndex = Interlocked.Increment(ref s_thunksCreated);
+                    Type[] paramTypes = new Type[parameters.Length + 1];
+                    paramTypes[0] = typeof(Func<object[], object>);
+
+                    StringBuilder thunkName = new StringBuilder();
+                    thunkName.Append("Thunk");
+                    thunkName.Append(thunkIndex);
+                    if (hasReturnValue)
+                    {
+                        thunkName.Append("ret_");
+                        thunkName.Append(returnType.Name);
+                    }
 
-                // load delegate
-                ilgen.Emit(OpCodes.Ldarg_0);
+                    for (int i = 0; i < parameters.Length; i++)
+                    {
+                        thunkName.Append("_");
+                        thunkName.Append(parameters[i].ParameterType.Name);
+                        paramTypes[i + 1] = parameters[i].ParameterType;
+                    }
 
-                // load array
-                ilgen.Emit(OpCodes.Ldloc, argArray);
+                    DynamicMethod dynamicThunkMethod = new DynamicMethod(thunkName.ToString(), returnType, paramTypes);
```

Comment:
Naive question, but I assume there's no character valid in a type name that's invalid in a method name, right?

Reply:
There is not. DynamicMethod's follow special rules, and the name of a dynamic method can be anything which is encodable in Utf8. As the file format of .NET is Utf8 based for names, any type name which can be converted into a string object is representable as Utf8. This implies that there is technically some risk here around the behavior of types which are encoded as invalid utf8, which are loadable by the runtime but not nameable via the reflection apis.